File: nginx_test/was-4-springboot/src/main/java/com/nginx/test/service/CacheService.java
package com.nginx.test.service;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.redis.core.StringRedisTemplate;
import org.springframework.stereotype.Service;

import java.time.Duration;
import java.util.HashMap;
import java.util.Map;

@Slf4j
@Service
@RequiredArgsConstructor
public class CacheService {

    private final StringRedisTemplate redisTemplate;

    public String get(String key) {
        try {
            String value = redisTemplate.opsForValue().get(key);
            log.debug("Cache GET: key={}, found={}", key, value != null);
            return value;
        } catch (Exception e) {
            log.error("Cache GET error: key={}, error={}", key, e.getMessage());
            return null;
        }
    }

    public void set(String key, String value, long ttlSeconds) {
        try {
            redisTemplate.opsForValue().set(key, value, Duration.ofSeconds(ttlSeconds));
            log.debug("Cache SET: key={}, ttl={}", key, ttlSeconds);
        } catch (Exception e) {
            log.error("Cache SET error: key={}, error={}", key, e.getMessage());
            throw e;
        }
    }

    public boolean delete(String key) {
        try {
            Boolean deleted = redisTemplate.delete(key);
            log.debug("Cache DELETE: key={}, deleted={}", key, deleted);
            return Boolean.TRUE.equals(deleted);
        } catch (Exception e) {
            log.error("Cache DELETE error: key={}, error={}", key, e.getMessage());
            return false;
        }
    }

    public Map<String, Object> healthCheck() {
        Map<String, Object> status = new HashMap<>();
        try {
            String pong = redisTemplate.getConnectionFactory().getConnection().ping();
            status.put("status", "PONG".equals(pong) ? "UP" : "DOWN");
            status.put("response", pong != null ? pong : "NO_RESPONSE");
        } catch (Exception e) {
            status.put("status", "DOWN");
            status.put("error", e.getMessage() != null ? e.getMessage() : "Unknown error");
        }
        return status;
    }
}

File: nginx_test/was-3-springboot/src/main/kotlin/com/nginx/test/service/CacheService.kt
package com.nginx.test.service

import io.micrometer.observation.annotation.Observed
import org.slf4j.LoggerFactory
import org.springframework.data.redis.core.StringRedisTemplate
import org.springframework.stereotype.Service
import java.time.Duration

@Service
@Observed(name = "cache.service")
class CacheService(
    private val redisTemplate: StringRedisTemplate
) {
    private val logger = LoggerFactory.getLogger(javaClass)

    fun get(key: String): String? {
        return try {
            val value = redisTemplate.opsForValue().get(key)
            logger.debug("Cache GET: key=$key, found=${value != null}")
            value
        } catch (e: Exception) {
            logger.error("Cache GET error: key=$key, error=${e.message}")
            null
        }
    }

    fun set(key: String, value: String, ttlSeconds: Long = 3600) {
        try {
            redisTemplate.opsForValue().set(key, value, Duration.ofSeconds(ttlSeconds))
            logger.debug("Cache SET: key=$key, ttl=$ttlSeconds")
        } catch (e: Exception) {
            logger.error("Cache SET error: key=$key, error=${e.message}")
            throw e
        }
    }

    fun delete(key: String): Boolean {
        return try {
            val deleted = redisTemplate.delete(key)
            logger.debug("Cache DELETE: key=$key, deleted=$deleted")
            deleted
        } catch (e: Exception) {
            logger.error("Cache DELETE error: key=$key, error=${e.message}")
            false
        }
    }

    fun healthCheck(): Map<String, Any> {
        return try {
            val pong = redisTemplate.connectionFactory?.connection?.ping()
            mapOf(
                "status" to if (pong == "PONG") "UP" else "DOWN",
                "response" to (pong ?: "NO_RESPONSE")
            )
        } catch (e: Exception) {
            mapOf(
                "status" to "DOWN",
                "error" to (e.message ?: "Unknown error")
            )
        }
    }
}
